presto-raptor/src/main/java/com/facebook/presto/raptor/storage/DeltaShardRewriter.java

```diff
@@ -0,0 +1,186 @@
+    {
+        if (rowsToDelete.isEmpty()) {
+            return ImmutableList.of();
+        }
+        // blockToDelete is LongArrayBlock
+        List<Long> columnIds = new ArrayList<>();
+        List<Type> columnTypes = new ArrayList<>();
+        columnIds.add(0L);
+        columnTypes.add(BIGINT);
+
+        // TODO: Under current implementation, one block can only hold INT_MAX many rows
+        //  which theoretically may not be enough to hold all rows from an ORC file.
+        // At this point, rowsToDelete couldn't be empty
+        oldDeltaDeleteShardUuid.ifPresent(oldDeltaDeleteShardUuid -> mergeToRowsToDelete(rowsToDelete, oldDeltaDeleteShardUuid));
+
+        if (rowsToDelete.cardinality() == getRowCount(oldShardUuid)) {
+            // Delete original file
+            return shardDeleteDelta(oldShardUuid, oldDeltaDeleteShardUuid, Optional.empty());
+        }
+
+        BlockBuilder blockBuilder = new LongArrayBlockBuilder(null, rowsToDelete.size());
+        for (int i = rowsToDelete.nextSetBit(0); i >= 0; i = rowsToDelete.nextSetBit(i + 1)) {
+            blockBuilder.writeLong(i);
+        }
+        Block blockToDelete = blockBuilder.build();
+        // TODO: a async call made by deletionExecutor, it calls into OrcPageSink which uses a different thread pool (i.e. commitExecutor)
+        //  Better use consistent thread pool management
+        StoragePageSink pageSink = orcStorageManager.createStoragePageSink(fileSystemContext, transactionId, bucketNumber, columnIds, columnTypes, true);
+        pageSink.appendPages(ImmutableList.of(new Page(blockToDelete)));
+        List<ShardInfo> shardInfos = getFutureValue(pageSink.commit());
+        // Guaranteed that shardInfos only has one element since we only call commit one time
+        ShardInfo newDeltaDeleteShard = Iterables.getOnlyElement(shardInfos);
+        return shardDeleteDelta(oldShardUuid, oldDeltaDeleteShardUuid, Optional.of(newDeltaDeleteShard));
+    }
+
+    // Note: This function will change rowsToDelete.
+    // Will merge the BitSet from oldDeltaDeleteShardUuid to rowsToDelete
+    // rowsToDelete and rowsDeleted must be mutually exclusive
+    private void mergeToRowsToDelete(BitSet rowsToDelete, UUID oldDeltaDeleteShardUuid)
+    {
+        Optional<BitSet> rowsDeleted = orcStorageManager.getRowsFromUuid(fileSystem, Optional.of(oldDeltaDeleteShardUuid));
+        BitSet verify = new BitSet();
+        verify.or(rowsToDelete);
+        verify.and(rowsDeleted.get());
+        if (verify.cardinality() != 0) {
+            throw new PrestoException(RAPTOR_ERROR, "rowsToDelete and rowsDeleted are not mutually exclusive");
+        }
+        if (rowsDeleted.isPresent()) {
+            rowsToDelete.or(rowsDeleted.get());
+        }
+    }
+
+    private int getRowCount(UUID oldShardUuid)
+    {
+        try (OrcDataSource dataSource = orcStorageManager.openShard(fileSystem, oldShardUuid, defaultReaderAttributes)) {
+            OrcReader reader = new OrcReader(
```

Comment:
This `new OrcReader` method seems to be used quite often; can we make it a helper in a util?